Implement a default for `jacobian_conjugate`

kellertuer opened the pull request:
This solves #20.

kellertuer committed:
Implement a default for `jacobian_conjugate`

fix usage of hat/vee, since they are already defined in ManifoldsBase

mateuszbaran commented on src/groups/product_group.jl:
```diff
@@ -344,7 +344,7 @@ function Base.show(
 
 function get_coordinates_lie!(
-    Pr𝔤::LieAlgebra{𝔽,Op,LieGroup{𝔽,Op,M}}, c, X, B::DefaultLieAlgebraOrthogonalBasis
+    Pr𝔤::LieAlgebra{𝔽,Op,LieGroup{𝔽,Op,M}}, c, X
```
Why isn't ` B` needed here?

kellertuer replied:
`get_coordinates_lie` is already the specific (Layer3) function for the `DefaultLieAlgebraOrthogonalBasis`, so this function does not need a basis any longer.

mateuszbaran replied:
There are variants of `DefaultLieAlgebraOrthogonalBasis` with real and complex coefficients and this information needs to be propagated here somehow (and I don't like passing just the number system, it's too much hassle).

mateuszbaran replied:
Anyway, `B` is even used inside so this function can't work without it.

kellertuer replied:
Hm, then this is overall quite inconsistent throughout LieGroups for now, I think I never passed (or dispatched on) basis in this Layer3 function.
But sure for other Layer-3-cases we do that, though most often with dispatching on the number system type I think.

mateuszbaran replied:
Yes, and IMO it was a bad idea to not pass the basis to layer-3 functions. I think this is a good time to stop doing it. I'm planning to redo ManifoldsBase dispatch accordingly when I have some spare time.

kellertuer replied:
I do not have a strong opinion on that, besides that it should be unified to some extend. Sure we can read it here, and even more if it is used inside, I seem to have missed it (and interesting that the tests work?) – and rework ManifoldsBase internally sounds also good then.

Will do that together with the code coverage for ValidationManifold after merging master back here – but not today. Tomorrow or Thursday probably.

mateuszbaran replied:
I did not have a strong opinion on that when it was introduced but I do have a strong opinion now after working with the current design a bit.

kellertuer replied:
Sure, sometimes it needs some time and using stuff to see what is better.

kellertuer replied:
But yeah, I merged master with the `ValidationLieGroup` but will add `B` than back in tomorrow and check that this is done everywhere for `get_vector_lie` and `get_coordinates_lie` then.

mateuszbaran commented on src/interface.jl:
```diff
@@ -645,9 +651,27 @@ end
+        get_vector!(𝔤, X, c, B)         # store ``X_j`` in X
+        diff_conjugate!(G, X, g, h, X) # compute the differential in-place
+        get_coordinates!(𝔤, c, X, B)   # compute its coordinates in B
+        J[:, j] .= c                    # store the result in J
```
```suggestion
        get_coordinates!(𝔤, view(J, :, j), X, B)   # compute its coordinates in B and store the result in J
```
This can work in one step

mateuszbaran commented:
Everything else looks good to me.

kellertuer committed:
Update src/interface.jl

Co-authored-by: Mateusz Baran <mateuszbaran89@gmail.com>

kellertuer committed:
Merge branch 'main' into kellertuer/jacobian_conjugate

# Conflicts:
#	src/interface.jl
#	test/LieGroupsTestSuite.jl/LieGroupsTestSuite.jl

kellertuer committed:
Merge branch 'kellertuer/jacobian_conjugate' of github.com:JuliaManifolds/LieGroups.jl into kellertuer/jacobian_conjugate

kellertuer committed:
Fix a hat/vee confusion

it is already defined in ManifoldsBase so let's extend that function with methods instead of coming up with our own.

kellertuer committed:
Fix another thing I missed in the merge.

kellertuer committed:
fix get_coordinate_lie

kellertuer committed:
add tests for jacobian_conjugate on the ValidationLieGroup

kellertuer commented:
Fixed the issue, was good, now all is unified – also brought validation jacobian_conjugate into the tests, so code cov also looks good. I think there is no need for an additional entry in the changelog, all is already mentioned.

kellertuer committed:
Add one manual allocation check.

kellertuer committed:
Update test/groups/test_special_euclidean_group.jl

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>
